### ISSUE TITLE
Feature request: Add ActiveSupport::Notification instrumentation for grape middlewares

### ISSUE THREAD

**Issue by pawelma** [ruby-grape/grape#2322]
# WHAT

Grape's support for `ActiveSupport::Notification` for performance monitoring is fantastic.

Is it feasible to expand by include instrumentation for middlewares invoked before `endpoint_run`?

# WHY

Tracing libraries, such as `NewRelic` agent, are [injecting themselves to gather transaction ](https://github.com/newrelic/newrelic-ruby-agent/blob/5c5d4d2c98c1392bf08910f7cd6bf2f846317d33/lib/new_relic/agent/instrumentation/grape/chain.rb#L9) data when they might just subscribe to the instrumentation details supplied straight from grape. This method works, however it may cause some maintenance issues while making structural modifications. It is possible that by permitting additional traces via `ActiveSupport::Notification`, all third-party tools will ultimately adopt.

`OpenTelemetry` [has subscribed](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/grape/lib/opentelemetry/instrumentation/grape/subscriber.rb) to `ActiveSupport::Notification`, but it is unable to track anything that happened within a middleware since there is no such trace provided.

Custom middleware for authentication is a simple example from docs for which there will be no trace:
https://github.com/ruby-grape/grape#register-custom-middleware-for-authentication

**Comment by dblock** [comment]
I don't see why not! Please contribute.

**Comment by pawelma** [comment]
I've looked into that yesterday, and It seems that it's not actually viable for grape to install instrumentation around the middlewares, as grape doesn't have control over them. `Stack::Middleware` is used to handle nice DSL and proxy to `Rack::Builder` in [build](https://github.com/ruby-grape/grape/blob/02fd374717a6b063e02dfb42f72f2dc80f9e0e29/lib/grape/middleware/stack.rb#L97-L103) method. 
Everything up to endpoint is handled by the Rack https://github.com/ruby-grape/grape/blob/02fd374717a6b063e02dfb42f72f2dc80f9e0e29/lib/grape/endpoint.rb#L316.

Workaround for app usage could be either create 'proxy' middleware and 'decorate' middleware that we want to track, or bind to `Rack.events` to get overall performance including middlewares.

Closing this feature request as implementing would probably require writing custom stack handler or injection into the middlewares that we're passing, which is probably not a good idea.
